### ISSUE TITLE
[Feature] Improved vehicle customization

### ISSUE THREAD

**Issue by FlorentP42** [Teardown-Issue-Tracker-Maintainers/Teardown-Issue-Tracker#334]
### Is your feature request related to a problem (if not, explain why you think this feature should be added)? Please describe.

So the vehicle entity is really handy and lets you easily control any kind of body with a nice follow cam. However, when you are trying to make very custom vehicles through mods (think flying plane, or mounted horse) there is several areas that cannot be fully edited and makes the final result less immersive:

1. You cannot fully override the vehicle controls displayed at bottom right of the screen when driving.
2. You cannot fully override the vehicles sounds.

### Ideal solution:

1. Add a new [API function / vehicle tag] that let you define the exact list of controls you want for your vehicle as a list of pair of key/control. If this list is defined at all, completely replace the default behaviour of the control display by the elements from this list.
2. Add a new [API function / vehicle tag] that let you define for each possible vehicle action the sound to play for that action, or no sound if left empty. For each sound defined this way it would override the default sound, for others it wiould keep the default sound.

### Good enough solution:

1. Add a new [API function / vehicle tag] that let you hide the default vehicle control panel. So we can draw our own modded one in its palce.
2. Add a new [API function / vehicle tag] that let you mute all sounds from the vehicle. So we can play our own custom sounds whenever we need to.

### Describe alternatives you've considered

1. Right now I could only add my custom controls above the default ones, which is not fun... I tried to draw mines over the default ones but as there is no z-level control in the UI the default controls always ended up drawn on top of mines so this is not a viable solution either...
2. I used the tag `sound="none"` on the vehicle entity, which actually removes most of the vehicle sounds and throws a bunch of error in the console because `none/whateverSound.ogg` does not exist, which is fine. But some sounds are not correctly overriden by it: for example I can still hear the gear change sounds...

**Comment by squarebracket-s** [comment]
Using a workaround, replacing the control panel of a vehicle is possible. By recreating the path to the data/ui* directory and then putting all files of data/ui* in your recreated ui folder, you can basically modify any UI part of Teardown (while the mod is loaded)
You could potentially even change vehicle sounds by recreating data/ui/vehicles*, but I'm not sure about that.

Doing this actually increases the size of your mod more than you actually need (because of other files you don't need, but they're required for the whole UI to work), so implementing this in the API would indeed be a better solution.

While it is not implemented officially, a modder could make a "universal" API for modifying the control panel of vehicles. This should be pretty easy, as the vehicle control panel render code puts all vehicle controls in one table, 'info' to be exact. (P.S.: code is in data/ui/hud.lua)

Thinking even further, such APIs could interfere with other mods that also change UI code.

Another workaround would be disabling the Teardown UI (set hud.hide or hud.disable to true), and then drawing yours where the panel would be. To avoid confusion, I would redraw the vehicle condition.

*I am referring to the files of Teardowns' directory.

**Comment by YuLun-bili** [comment]
Problem with replacing vanilla scripts in `data` folder is these scripts are mostly mixed, `#include`-ing other scripts thus got tied up, so one have to (as you've said) replace almost all. At the same time, you cannot assume player would not use a mod that need to replace some of these scripts, therefore would cause a script replacement conflict and cause issues to any mods trying to replace and leaving only one working. So currently the best option isn't to replace scripts but simply make a custom script or don't add that feature to mod (which is sad).

**Comment by FlorentP42** [comment]
> Using a workaround, replacing the control panel of a vehicle is possible. By recreating the path to the data/ui* directory and then putting all files of data/ui* in your recreated ui folder, you can basically modify any UI part of Teardown (while the mod is loaded) You could potentially even change vehicle sounds by recreating data/ui/vehicles*, but I'm not sure about that.
> 
> Doing this actually increases the size of your mod more than you actually need (because of other files you don't need, but they're required for the whole UI to work), so implementing this in the API would indeed be a better solution.
> 
> While it is not implemented officially, a modder could make a "universal" API for modifying the control panel of vehicles. This should be pretty easy, as the vehicle control panel render code puts all vehicle controls in one table, 'info' to be exact. (P.S.: code is in data/ui/hud.lua)
> 
> Thinking even further, such APIs could interfere with other mods that also change UI code.
> 
> Another workaround would be disabling the Teardown UI (set hud.hide or hud.disable to true), and then drawing yours where the panel would be. To avoid confusion, I would redraw the vehicle condition.
> 
> *I am referring to the files of Teardowns' directory.

Yes, you approach seems more fitting for global mods that impact all vehicles and bring new features for all of them. What I am looking for here is a way to customize one single vehicle alone, only by changing a few tags in the prefab xml of said vehicle ;)

I mean yes I could override the main vehicle scripts myself to add that feature and then use it to customize only one, but as [YuLun](https://github.com/YuLun-bili) said it would make compatibility with other potential mods affectig vehicle really hard... So I would prefer not to do that :p

**Comment by Thomasims** [comment]
You don't need to override the base game UI files just to suppress the vehicle keybinds/health UI, calling `SetBool("game.vehicle.interactive", false)` repeatedly in `tick()` (while the player is in your vehicle) should be sufficient.

**Comment by squarebracket-s** [comment]
Just tested it. It is indeed the easiest way to draw your own controls for a vehicle. Only thing is, if you open the pause menu a part of the vehicle UI shows for 1 frame (obviously, since `tick()` is not called when the pause menu is open). I don't think this will annoy anyone though.

**Comment by FlorentP42** [comment]
> You don't need to override the base game UI files just to suppress the vehicle keybinds/health UI, calling `SetBool("game.vehicle.interactive", false)` repeatedly in `tick()` (while the player is in your vehicle) should be sufficient.

This would be a nice workaround for the UI, what about the sounds though?

**Comment by YuLun-bili** [comment]
Cannot change gear switching or brake sound if you still want to drive it normally, a bit annoying.
The tag `passive` does makes a vehicle have no gear sound but it would disable steering and make vehicle drive very slow.

**Comment by FlorentP42** [comment]
> Cannot change gear switching or brake sound if you still want to drive it normally, a bit annoying. The tag `passive` does makes a vehicle have no gear sound but it would disable steering and make vehicle drive very slow.

Actually that might be a good enough solution for the case I have currently (as I was completely overriding the controls with my own anyway I don't care if there is no native controls). But what could be annoying is if it was actively limiting the whole vehicle speed no matter how I try to move it through custom script though...

**Comment by FlorentP42** [comment]
> Cannot change gear switching or brake sound if you still want to drive it normally, a bit annoying. The tag `passive` does makes a vehicle have no gear sound but it would disable steering and make vehicle drive very slow.

Well I finally remembered to test this and it does cut the gear sound but also adds a grinding sound on the spacebar action, which is also bad for my use case XD
